src/bssl.rs

```diff
@@ -12,12 +12,62 @@
 // OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF OR IN
 // CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
 
-use {c, error};
+pub use self::ffi_error::ErrorCode;
 
-pub fn map_result(bssl_result: c::int) -> Result<(), error::Unspecified> {
-    match bssl_result {
-        1 => Ok(()),
-        _ => Err(error::Unspecified),
+/// Many foreign functions return a `c::int` indicating whether an error occurred. This module
+/// wraps that error code in a `#[repr(transparent)]` newtype which can be converted into a
+/// `Result`.
+mod ffi_error {
+    use core::marker::PhantomData;
+    use {c, error};
+
+    /// The meaning of the error code returned by a foreign function.
+    pub trait Semantics {
+        /// Returns true if a given error code indicates success.
+        fn is_success(code: c::int) -> bool;
+    }
+
+    /// The semantics used by most BoringSSL functions:
+    ///
+    /// * `1` indicates success.
+    /// * `0` indicates failure.
+    #[derive(Debug, Clone, Copy)]
+    pub enum BoringSsl {}
+
+    impl Semantics for BoringSsl {
+        fn is_success(code: c::int) -> bool {
+            match code {
+                1 => true,
+                c => {
+                    // Be pedantic in debug mode to catch some cases where function signatures are
+                    // wrongly defined.
+                    debug_assert_eq!(c, 0, "BoringSSL functions should return 0 or 1");
+                    false
+                }
+            }
+        }
+    }
+
+    /// An error code returned by a foreign function.
+    ///
+    /// The template parameter indicates the semantics of the returned value.
+    #[must_use]
+    #[repr(transparent)]
+    #[derive(Debug, Clone, Copy)]
+    pub struct ErrorCode<S = BoringSsl> {
```

Comment:
I don't think `ErrorCode` is a very good name for this boolean flag. I think `Result` would make more sense since BoringSSL's boolean results are basically `Result<(), ()>`s.

Reply:
Agreed. I avoided it initially since `modname::Result` is usually a type alias for `std::result::Result` with the error type defined.